File: sdk/python/payclaw/verify.py
import time
import requests

# keccak256("Transfer(address,address,uint256)")
TRANSFER_TOPIC = "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef"


def _rpc(rpc_url: str, method: str, params: list, retries: int = 3):
    payload = {"jsonrpc": "2.0", "id": 1, "method": method, "params": params}
    delay = 1.0
    last_err = None
    for _ in range(retries):
        try:
            resp = requests.post(rpc_url, json=payload, timeout=10)
            resp.raise_for_status()
            data = resp.json()
            if "error" in data:
                raise ValueError(f"RPC error: {data['error']}")
            return data.get("result")
        except Exception as e:
            last_err = e
            time.sleep(delay)
            delay *= 2
    raise ConnectionError(f"RPC failed after {retries} retries: {last_err}")


def _decode_address(padded: str) -> str:
    return "0x" + padded[-40:]


def verify_payment(tx_hash: str, config, nonce_cache) -> tuple[bool, str]:
    if nonce_cache.is_used(tx_hash):
        return False, "tx already used"

    chain_id_hex = _rpc(config.rpc_url, "eth_chainId", [])
    actual_chain_id = int(chain_id_hex, 16)
    if actual_chain_id != config.chain_id:
        return False, f"chain mismatch: expected {config.chain_id}, got {actual_chain_id}"

    tx = _rpc(config.rpc_url, "eth_getTransactionByHash", [tx_hash])
    if tx is None:
        return False, "tx not found"

    # USDC is ERC-20 — the amount is NOT in tx.value (that is always 0 for token transfers).
    # The real amount is in the Transfer event log of the receipt.
    receipt = _rpc(config.rpc_url, "eth_getTransactionReceipt", [tx_hash])
    if receipt is None:
        return False, "tx receipt not found"

    # Reject reverted transactions — logs are stripped on revert in EVM, but check anyway
    if receipt.get("status") != "0x1":
        return False, "tx reverted (status != 0x1)"

    usdc_log = None
    for log in receipt.get("logs", []):
        topics = log.get("topics", [])
        if (
            len(topics) >= 3
            and topics[0].lower() == TRANSFER_TOPIC
            and log.get("address", "").lower() == config.usdc_address.lower()
            and _decode_address(topics[2]).lower() == config.wallet_address.lower()
        ):
            usdc_log = log
            break

    if usdc_log is None:
        return False, "no USDC transfer to recipient found in tx"

    try:
        value = int(usdc_log.get("data", "0x0"), 16)
    except ValueError:
        return False, "malformed transfer amount in log data"
    if value < config.price_units:
        return False, f"insufficient USDC: got {value} units, need {config.price_units}"

    block = _rpc(config.rpc_url, "eth_getBlockByNumber", [tx.get("blockNumber"), False])
    if block is None:
        return False, "block not found"

    block_ts = int(block["timestamp"], 16)
    age = abs(int(time.time()) - block_ts)
    if age > config.freshness_seconds:
        return False, f"tx too old: {age}s (max {config.freshness_seconds}s)"

    if not nonce_cache.mark_used(tx_hash):
        return False, "tx already used (race condition)"

    return True, "ok"

File: sdk/js/src/verify.ts
import type { ResolvedConfig, VerifyResult } from "./types.js";

// keccak256("Transfer(address,address,uint256)")
const TRANSFER_TOPIC = "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef";

async function rpc(url: string, method: string, params: unknown[], retries = 3): Promise<unknown> {
  let delay = 1000;
  let lastErr: unknown;
  for (let i = 0; i < retries; i++) {
    try {
      const res = await fetch(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ jsonrpc: "2.0", id: 1, method, params }),
      });
      if (!res.ok) throw new Error(`HTTP ${res.status}`);
      const data = (await res.json()) as { result?: unknown; error?: unknown };
      if (data.error) throw new Error(`RPC error: ${JSON.stringify(data.error)}`);
      return data.result ?? null;
    } catch (e) {
      lastErr = e;
      if (i < retries - 1) await sleep(delay);
      delay *= 2;
    }
  }
  throw new Error(`RPC failed after ${retries} retries: ${lastErr}`);
}

function sleep(ms: number) {
  return new Promise((r) => setTimeout(r, ms));
}

function decodeAddress(padded: string): string {
  return "0x" + padded.slice(-40);
}

export async function verifyPayment(txHash: string, config: ResolvedConfig): Promise<VerifyResult> {
  if (await config.nonceStore.has(txHash)) {
    return { ok: false, reason: "tx already used" };
  }

  const tx = (await rpc(config.rpcUrl, "eth_getTransactionByHash", [txHash])) as Record<string, string> | null;
  if (!tx) return { ok: false, reason: "tx not found" };

  // USDC is ERC-20 — amount is in Transfer event logs, not tx.value (which is 0).
  const receipt = (await rpc(config.rpcUrl, "eth_getTransactionReceipt", [txHash])) as {
    logs: Array<{ address: string; topics: string[]; data: string }>;
  } | null;
  if (!receipt) return { ok: false, reason: "tx receipt not found" };

  const usdcLog = receipt.logs.find(
    (log) =>
      log.topics.length >= 3 &&
      log.topics[0].toLowerCase() === TRANSFER_TOPIC &&
      log.address.toLowerCase() === config.usdcAddress.toLowerCase() &&
      decodeAddress(log.topics[2]).toLowerCase() === config.walletAddress.toLowerCase()
  );

  if (!usdcLog) return { ok: false, reason: "no USDC transfer to recipient found in tx" };

  const value = BigInt(usdcLog.data);
  if (value < config.priceUnits) {
    return { ok: false, reason: `insufficient USDC: got ${value} units, need ${config.priceUnits}` };
  }

  const block = (await rpc(config.rpcUrl, "eth_getBlockByNumber", [tx.blockNumber, false])) as {
    timestamp: string;
  } | null;
  if (!block) return { ok: false, reason: "block not found" };

  const blockTs = Number(BigInt(block.timestamp));
  const now = Math.floor(Date.now() / 1000);
  const age = Math.abs(now - blockTs);
  if (age > config.freshnessSeconds) {
    return { ok: false, reason: `tx too old: ${age}s (max ${config.freshnessSeconds}s)` };
  }

  const marked = await config.nonceStore.set(txHash, config.nonceCacheTtl);
  if (!marked) return { ok: false, reason: "tx already used (race condition)" };

  return { ok: true };
}
